### PR TITLE
Add cookie-based-affinity-distinct-name annotation

### LINKED ISSUE #772: Enhance annotation list of AGIC
## Is your feature request related to a problem? Please describe
Greetings! I am not sure whether this is a feature request, but I decided to post this here.

I think that it would be beneficial to support the override of the *Affinity cookie name* property in Application Gateway, as there are some applications which would not use the default affinity cookie name and would require a custom one.

In addition to this, it would be a good idea to also support the overwrite of the incoming HTTP host header, as the domain of some backends can be different from the DNS of the Application Gateway and the backend would rely on a specific host header to resolve to the correct endpoint.

Both these properties are valid for the application that I am trying to use AGIC with and would be very grateful if you consider if what is required is possible.

## Describe the solution you'd like
As a solution, I imagine having both properties added to the AGIC annotation list.

### Affinity cookie name
For this property it should only be working if ```appgw.ingress.kubernetes.io/cookie-based-affinity: "true"```

#### Usage
```yaml 
appgw.ingress.kubernetes.io/affinity-cookie-name: "value"
```
#### Example
```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: go-server-ingress-affinity
  namespace: test-ag
  annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    appgw.ingress.kubernetes.io/cookie-based-affinity: "true"
    appgw.ingress.kubernetes.io/affinity-cookie-name: "test"
spec:
  rules:
  - http:
      paths:
      - path: /hello/
        backend:
          serviceName: go-server-service
          servicePort: 80
```
#### Usage
To overwrite the HTTP host header with host from backend target:
```yaml 
appgw.ingress.kubernetes.io/overwrite-http-header-from-backend-target: "true"
```
To overwrite with specific domain:
```yaml 
appgw.ingress.kubernetes.io/overwrite-http-header: "example.com"
```
#### Example 1
```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: go-server-ingress-affinity
  namespace: test-ag
  annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    appgw.ingress.kubernetes.io/overwrite-http-header-from-backend-target: "true"
spec:
  rules:
  - http:
      paths:
      - path: /hello/
        backend:
          serviceName: go-server-service
          servicePort: 80
```
#### Example 2
```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: go-server-ingress-affinity
  namespace: test-ag
  annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    appgw.ingress.kubernetes.io/overwrite-http-header: "example.com"
spec:
  rules:
  - http:
      paths:
      - path: /hello/
        backend:
          serviceName: go-server-service
          servicePort: 80
```
## Conclusion
Please let me know if I this functionality is already present and I have missed it for some reason.
If not please consider whether this is possible and how much time would it cost.
Thanks!

### PR DESCRIPTION
**Pull request by rbickel**
## Checklist
- [X] The title of the PR is clear and informative
- [X] If applicable, the changes made in the PR have proper test coverage
- [X] Issues addressed by the PR are mentioned in the description followed by `Fixes`.

## Description

Generate a unique affinity cookie name when the annotation `cookie-based-affinity-distinct-name` is set to `true` in an ingress. This maintains affinity when using path based routing for different backends..

Adding the following annotations will ensure every backend has a different affinity cookie:
```yaml
    appgw.ingress.kubernetes.io/cookie-based-affinity: "true"
    appgw.ingress.kubernetes.io/cookie-based-affinity-distinct-name: "true"
```

## Fixes

#772 Enhance annotation list of AGIC

### CODE REVIEW

**Comment by akshaysngupta** [comment]
@rbickel Thanks for the PR.

How is the application using the the cookie at the backend ? Would backend application need to know the cookie name beforehand ?

**Comment by rbickel** [comment]
The backend doesn't use the cookie. It is only for application gateway to route the request to the proper backend. If you have 2 different routing path on your ingress, when moving from a route to another, it overwrites the cookie with the affinity of the other path. You need different cookie name to ensure when you navigate back you still hit the same backend.
Affinity cookies in Application Gateway are just a hash of the pod IP, so there isn't any usage of this on the backend
e.g. with below ingress. When navigating to one path to the other, the default cookie `ApplicationGatewayAffinity` is overridden. When navigating back to the other path, that may have multiple pods as replicas, the affinity is lost :
```yaml
apiVersion: extensions/v1beta1
kind: Ingress
metadata:
  name: go-server-ingress-affinity
  namespace: test-ag
  annotations:
    kubernetes.io/ingress.class: azure/application-gateway
    appgw.ingress.kubernetes.io/cookie-based-affinity: "true"
    appgw.ingress.kubernetes.io/cookie-based-affinity-distinct-name: "true"
spec:
  rules:
  - http:
      paths:
      - path: /affinity1/
        backend:
          serviceName: affinity-service
          servicePort: 80
      - path: /affinity2/
        backend:
          serviceName: other-affinity-service
          servicePort: 80          
```

**Comment by rbickel** [comment]
@akshaysngupta  We could prefix the cookie name with `appgw-affinity-` and then the backend hash, so we provide some hints to the browser what is the cookie for ? However I didn't want to leak appgw backend information back to the browser, reason of why I am using a hash.

File: pkg/appgw/backendhttpsettings.go [hunk 1]
```diff
@@ -275,6 +275,12 @@ func (c *appGwConfigBuilder) generateHTTPSettings(backendID backendIdentifier, p
 		c.recorder.Event(backendID.Ingress, v1.EventTypeWarning, events.ReasonInvalidAnnotation, err.Error())
 	}
 
+	if distinctName, err := annotations.IsCookieBasedAffinityDistinctName(backendID.Ingress); err == nil && distinctName {
+		httpSettings.AffinityCookieName = to.StringPtr(backendID.serviceFullNameHash())
```
Comment (akshaysngupta) [inline pkg/appgw/backendhttpsettings.go:279]: What do you think about appending `appgw-affinity` to the cookie name ?

File: functional_tests/cookie_name.json [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+                    ]
+                }
+            }
+        ],
+        "backendHttpSettingsCollection": [
+            {
+                "id": "/subscriptions/--subscription--/resourceGroups/--resource-group--/providers/Microsoft.Network/applicationGateways/--app-gw-name--/backendHttpSettingsCollection/bp---namespace---hello-world-a-80-80---name--SlashNothingSlashSomething",
+                "name": "bp---namespace---hello-world-a-80-80---name--SlashNothingSlashSomething",
+                "properties": {
+                    "affinityCookieName": "3362c921cf90d228a7cd586ac3d72008",
```
Comment (akshaysngupta) [inline functional_tests/cookie_name.json:51]: this test is failing

File: functional_tests/cookie_name.json [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+                    },
+                    "protocol": "Http",
+                    "requestTimeout": 30
+                }
+            },
+            {
+                "id": "/subscriptions/--subscription--/resourceGroups/--resource-group--/providers/Microsoft.Network/applicationGateways/--app-gw-name--/backendHttpSettingsCollection/bp---namespace---hello-world-b-80-80---name--SlashNothingSlashSomething",
+                "name": "bp---namespace---hello-world-b-80-80---name--SlashNothingSlashSomething",
+                "properties": {
+                    "affinityCookieName": "ef38541db88f854ddb5b951dd04b333d",
```
Comment (akshaysngupta) [inline functional_tests/cookie_name.json:66]: here also `appgw-affinity-`

**Review by akshaysngupta — APPROVED** [review]

**Pull request merged**